### ISSUE TITLE
BertDataBunch' object has no attribute 'model_type'

### ISSUE THREAD

**Issue by JRatschat** [kaushaltrivedi/fast-bert#220]
I have been following the tutorials concerning Fast-Bert:
https://pypi.org/project/fast-bert/
https://www.kaggle.com/c/jigsaw-unintended-bias-in-toxicity-classification/discussion/92668

My goal is to do binary text classifictation. Therefore, my label.csv has only two labels and I set multi_label to False.

When executing `BertLearner.from_pretrained_model`, I am receiving the following error:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-240-ef4cead1d6f0> in <module>
     16                                             loss_scale = args['loss_scale'],
     17                                             multi_gpu = True,
---> 18                                             multi_label = False)

~/.local/lib/python3.6/site-packages/fast_bert/learner_cls.py in from_pretrained_model(dataBunch, pretrained_path, output_dir, metrics, device, logger, finetuned_wgts_path, multi_gpu, is_fp16, loss_scale, warmup_steps, fp16_opt_level, grad_accumulation_steps, multi_label, max_grad_norm, adam_epsilon, logging_steps, freeze_transformer_layers)
    131         model_state_dict = None
    132 
--> 133         model_type = dataBunch.model_type
    134 
    135         if torch.cuda.is_available():

AttributeError: 'BertDataBunch' object has no attribute 'model_type'
```

What I have tried so far is including  `model_type = 'bert'` to the `BertDataBunch` command. This has not helped so far. I am quite sure that my .csv's are in the right format, but of course, this could also be one source of the problem. PATH and imported modules should be fine.

Attached you find my code:

```
from pytorch_pretrained_bert.tokenization import BertTokenizer
from fast_bert.data import BertDataBunch

# Default args. If GPU runs out of memory while training, decrease training
# batch size
args = Box({
    "run_text": "tweet sentiment",
    "task_name": "Tweet Sentiment",
    "max_seq_length": 512,
    "do_lower_case": True,
    "train_batch_size": 8,
    "learning_rate": 6e-5,
    "num_train_epochs": 12.0,
    "warmup_proportion": 0.002,
    "local_rank": -1,
    "gradient_accumulation_steps": 1,
    "fp16": True,
    "loss_scale": 128
})
```

```
device = torch.device('cuda')

# check if multiple GPUs are available
if torch.cuda.device_count() > 1:
    multi_gpu = True
else:
    multi_gpu = False

# The tokenizer object is used to split the text into tokens used in training
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case = args['do_lower_case'])
    
# Databunch    
databunch = BertDataBunch(DATA_PATH, LABEL_PATH,
                          tokenizer = tokenizer,
                          train_file = 'X_train.csv', 
                          val_file = 'X_test.csv', 
                          label_file = 'label.csv',
                          text_col = 'text',
                          label_col = 'label',
                          bs = args['train_batch_size'], 
                          maxlen = args['max_seq_length'], 
                          multi_gpu = True, 
                          multi_label = False,
                          model_type = 'bert')

databunch.save()
num_labels = len(databunch.labels)
num_labels
```

```
# Set logger
import logging
import sys

logfile = str(LOG_PATH/'log-{}-{}.txt'.format(run_start_time, args["run_text"]))

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(name)s -   %(message)s',
    datefmt='%m/%d/%Y %H:%M:%S',
    handlers=[
        logging.FileHandler(logfile),
        logging.StreamHandler(sys.stdout)
    ])

logger = logging.getLogger()
logger.info(args)
```

When executing this field, the error happens:

```
from fast_bert.learner_cls import BertLearner
from fast_bert.metrics import accuracy

# Choose the metrics used for the error function in training
metrics = []
metrics.append({'name': 'accuracy', 'function': accuracy})

learner = BertLearner.from_pretrained_model(databunch, 
                                            pretrained_path = "bert-base-uncased", 
                                            metrics = metrics, 
                                            device = device,
                                            logger = logger, 
                                            output_dir = OUTPUT_DIR,
                                            finetuned_wgts_path = None, 
                                            is_fp16 = args['fp16'], 
                                            loss_scale = args['loss_scale'],
                                            multi_gpu = True,
                                            multi_label = False)
```

Thank you for your help!

**Comment by lingdoc** [comment]
What is the reason for the line:
```
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case = args['do_lower_case'])
```
FastBert uses "fast" versions of named tokenizers from HuggingFace (https://huggingface.co/models?search=bert) for the base models, which use the same defaults that you are setting here. (see https://github.com/kaushaltrivedi/fast-bert/blob/master/fast_bert/data_cls.py)

So if you replace `tokenizer` with `bert-base-uncased` in your databunch call, does that work? i.e.:
```
databunch = BertDataBunch(DATA_PATH, LABEL_PATH,
                          tokenizer = 'bert-base-uncased',
                          train_file = 'X_train.csv', 
                          val_file = 'X_test.csv', 
                          label_file = 'label.csv',
                          text_col = 'text',
                          label_col = 'label',
                          bs = args['train_batch_size'], 
                          maxlen = args['max_seq_length'], 
                          multi_gpu = True, 
                          multi_label = False,
                          model_type = 'bert')
```

**Comment by JRatschat** [comment]
@lingdoc Thank you for your answer. I have switched `tokenizer` with `bert-base-uncased` but the error is still present.

**Comment by lingdoc** [comment]
That's very strange. The error `'BertDataBunch' object has no attribute 'model_type'` basically means there is no 'model_type' variable in your databunch object. Are you trying to load the databunch from a file and run your training on that? Maybe something is up with how it gets serialized.

In my code it works without calling `databunch.save()`, but I store the databunch variable in memory before running training on top of that. maybe try to create the two variables in succession in the same script, something along the lines of:
```
from pytorch_pretrained_bert.tokenization import BertTokenizer
from fast_bert.data import BertDataBunch
from fast_bert.learner_cls import BertLearner
from fast_bert.metrics import accuracy

databunch = BertDataBunch(DATA_PATH, LABEL_PATH,
                          tokenizer = 'bert-base-uncased',
                          train_file = 'X_train.csv', 
                          val_file = 'X_test.csv', 
                          label_file = 'label.csv',
                          text_col = 'text',
                          label_col = 'label',
                          bs = args['train_batch_size'], 
                          maxlen = args['max_seq_length'], 
                          multi_gpu = True, 
                          multi_label = False,
                          model_type = 'bert')

learner = BertLearner.from_pretrained_model(databunch, 
                                            pretrained_path = "bert-base-uncased", 
                                            metrics = metrics, 
                                            device = device,
                                            logger = logger, 
                                            output_dir = OUTPUT_DIR,
                                            finetuned_wgts_path = None, 
                                            is_fp16 = args['fp16'], 
                                            loss_scale = args['loss_scale'],
                                            multi_gpu = True,
                                            multi_label = False)
```

**Comment by JRatschat** [comment]
@lingdoc thanks again for your help. No, I just used the save command to control if the `BertDataBunch` does anything... What do you mean with 

> I store the databunch variable in memory before running training on top of that?

That's what I have done as well, or not?

I have tried your approach, but I still get the same error :/ I don't really get it, because in the `BertDataBunch` command, I specify the `model_type` as `"bert"`.

**Comment by lingdoc** [comment]
yeah, that's weird. I don't know what's going on. the error message means that the DataBunch which gets stored in the `databunch` variable doesn't have a 'model_type' attribute. But I don't know why this is happening to you.

The only other trouble shooting step I can recommend is to compare the respective databunch variables in your code with those in the `README.md` file at the head of this repo. Some of your attributes don't quite seem to match the tutorial documentation. (`bs` should probably be `batch_size_per_gpu`, for example)

**Comment by JRatschat** [comment]
I finally found the issue. I loaded `BertDataBunch` from from `fast_bert.data import BertDataBunch` instead of  `from fast_bert.data_cls import BertDataBunch`. Now everything works. Thank you very much @lingdoc and sorry that I did not spot this mistake earlier! Problem was probably that I was following an outdated tutorial. Your comment 

> Some of your attributes don't quite seem to match the tutorial documentation

led me to the mistake. So thank you again!
